Mathlib/RingTheory/AdicCompletion/Functoriality.lean

```diff
@@ -0,0 +1,358 @@
+/-
+Copyright (c) 2024 Judith Ludwig, Christian Merten. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Judith Ludwig, Christian Merten
+-/
+import Mathlib.RingTheory.AdicCompletion.Basic
+import Mathlib.RingTheory.AdicCompletion.Algebra
+import Mathlib.Algebra.DirectSum.Basic
+
+/-!
+# Functoriality of adic completions
+
+In this file we establish functorial properties of the adic completion.
+
+## Main definitions
+
+- `LinearMap.adicCauchy I f`: the by `f` induced linear map on `I`-adic cauchy sequences
+- `LinearMap.adicCompletion I f`: the by `f` induced linear map on `I`-adic completions
+
+## Main results
+
+- `sumEquivOfFintype`: adic completion commutes with finite sums
+- `piEquivOfFintype`: adic completion commutes with finite products
+
+-/
+
+variable {R : Type*} [CommRing R] (I : Ideal R)
+
+namespace LinearMap
+
+variable {M : Type*} [AddCommGroup M] [Module R M]
+variable {N : Type*} [AddCommGroup N] [Module R N]
+variable {P : Type*} [AddCommGroup P] [Module R P]
+variable {T : Type*} [AddCommGroup T] [Module (AdicCompletion I R) T]
+
+attribute [-simp] smul_eq_mul Algebra.id.smul_eq_mul
+
+/-- `R`-linear version of `reduceModIdeal`. -/
+private def reduceModIdealAux (f : M →ₗ[R] N) :
+    M ⧸ (I • ⊤ : Submodule R M) →ₗ[R] N ⧸ (I • ⊤ : Submodule R N) :=
+  Submodule.mapQ (I • ⊤ : Submodule R M) (I • ⊤ : Submodule R N) f
+    (fun x hx ↦ by
+      refine Submodule.smul_induction_on hx (fun r hr x _ ↦ ?_) (fun x y hx hy ↦ ?_)
+      · simp [Submodule.smul_mem_smul hr Submodule.mem_top]
+      · simp [Submodule.add_mem _ hx hy])
+
+@[simp]
+private theorem reduceModIdealAux_apply (f : M →ₗ[R] N) (x : M) :
+    (f.reduceModIdealAux I) (Submodule.Quotient.mk (p := (I • ⊤ : Submodule R M)) x) =
+      Submodule.Quotient.mk (p := (I • ⊤ : Submodule R N)) (f x) :=
+  rfl
+
+/-- The induced linear map on the quotients mod `I • ⊤`. -/
+def reduceModIdeal (f : M →ₗ[R] N) :
+    M ⧸ (I • ⊤ : Submodule R M) →ₗ[R ⧸ (I • ⊤ : Ideal R)] N ⧸ (I • ⊤ : Submodule R N) where
```

Comment:
Why `R ⧸ (I • ⊤ : Ideal R)` instead of `R ⧸ I`?

Reply:
Because I thought it is more convenient to work with, because of the definition of `AdicCompletion I R`. Turns out, it's also doable with the canonical spelling, so I changed it.